### DISCUSSION TITLE
How to use the `state< S, R... >` metarule.

### DISCUSSION THREAD

**Discussion by tylercecil** [taocpp/PEGTL#9] [Q&A] [answered]
I'm having trouble understanding the `state` metarule, and there aren't currently any examples that use it (please correct me if I'm wrong, and point me to the example). Just some basic questions:
- What constructors must be available for my state type?
- What should the signature of `success()` be?
  I think a small example would really help me out, if possible.

**Comment by d-frey** [comment, marked as answer]
You can have a look at the [Switching-Style](https://github.com/ColinH/PEGTL/wiki/Switching-Style#intrusive-switching) documentation page, there we also mention `pegtl/contrib/raw_string.hh` as an example.

In the example, you can also find the basic signature for both the constructor as well as the `success()` method:

```
template< typename Input, typename ... States >
raw_string_state( const Input &, States && ... )
{ }

template< typename Input, typename ... States >
void success( const Input &, States && ... ) const
{ }
```

This applies if you want to basically ignore all the (outer) states (which might as well be no states). In most cases, however, you want to apply the collected data of your state class to the outer state.

Is this sufficient? (We might want to extend the example-section in the future anyways, but don't hold your breath)

**Comment by tylercecil** [comment]
Thanks @d-frey, that's really all that I needed! More examples in the future might be helpful, but for now I think I'll be ok.
